packages/cursorless-vscode/src/ide/vscode/VscodeFileSystem.ts

```diff
@@ -1,52 +1,15 @@
-import {
-  Disposable,
-  FileSystem,
-  PathChangeListener,
-  walkFiles,
-} from "@cursorless/common";
-import { stat } from "fs/promises";
-import { max } from "lodash";
+import { Disposable, FileSystem, PathChangeListener } from "@cursorless/common";
+import { RelativePattern, Uri, workspace } from "vscode";
 
 export class VscodeFileSystem implements FileSystem {
   watchDir(path: string, onDidChange: PathChangeListener): Disposable {
-    // Just poll for now; we can take advantage of VSCode's sophisticated
-    // watcher later. Note that we would need to do a version check, as VSCode
-    // file watcher is only available in more recent versions of VSCode.
-    return new PollingFileSystemWatcher(path, onDidChange);
-  }
-}
-
-const CHECK_INTERVAL_MS = 1000;
-
-class PollingFileSystemWatcher implements Disposable {
-  private maxMtimeMs: number = -1;
-  private timer: NodeJS.Timer;
-
-  constructor(
-    private readonly path: string,
-    private readonly onDidChange: PathChangeListener,
-  ) {
-    this.checkForChanges = this.checkForChanges.bind(this);
-    this.timer = setInterval(this.checkForChanges, CHECK_INTERVAL_MS);
-  }
-
-  private async checkForChanges() {
-    const paths = await walkFiles(this.path);
-
-    const maxMtime =
-      max(
-        (await Promise.all(paths.map((file) => stat(file)))).map(
-          (stat) => stat.mtimeMs,
-        ),
-      ) ?? 0;
-
-    if (maxMtime > this.maxMtimeMs) {
-      this.maxMtimeMs = maxMtime;
-      this.onDidChange();
-    }
-  }
-
-  dispose() {
-    clearInterval(this.timer);
+    // FIXME: Support globs?
+    const watcher = workspace.createFileSystemWatcher(
+      new RelativePattern(Uri.file(path), "**"),
+    );
+    watcher.onDidChange(onDidChange);
```

Comment:
We should probably handle these disposables as well

https://github.com/cursorless-dev/cursorless/blob/e2198dbd50faf9a38fe143916a285356ed14ef68/packages/cursorless-vscode/src/ide/vscode/hats/VscodeHatRenderer.ts#L419-L425

Reply:
I couldn't find anything definitive so I would err on the side of caution.

Reply:
fwiw it could also backfire, because it's possible disposing one of these `onFoo` methods after disposing the main watcher could cause an error to be thrown. I wonder if we should at least dispose the watcher after the `onFoo` methods 🤔

Reply:
k just went spelunking through VSCode code base, and the code uses an [`Emitter`](https://github.com/microsoft/vscode/blob/9e9c4e4bf713e143063414c59c8056b113f0263b/src/vs/base/common/event.ts#L946) for each event type (change / remove / add) and [disposes of its emitters](https://github.com/microsoft/vscode/blob/9e9c4e4bf713e143063414c59c8056b113f0263b/src/vs/workbench/api/common/extHostFileSystemEventService.ts#L104). From what I can tell there's no need to individually remove listeners when the emitter is disposed, and in the VSCode git extension, they only [dispose of the watcher](https://github.com/microsoft/vscode/blob/9e9c4e4bf713e143063414c59c8056b113f0263b/extensions/git/src/watch.ts#L19).

So long story short, we don't need to dispose of the watchers, and in fact I worry it could cause problems if we do dispose of them after the emitter is disposed

Reply:
Assuming you have no issues with my conclusions here, I can fix your code while I'm here

Reply:
Go ahead